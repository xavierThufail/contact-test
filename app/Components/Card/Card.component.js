import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { View, Text, TextInput, TouchableOpacity, StyleSheet, Image, ActivityIndicator, Button } from 'react-native';

import Icon from '../Icon/Icon.component';
import allAction from '../../Store/actions';
import styles from './Card.style';
import config from './Card.config';
import useCard from './Card.hook';

const renderImage = ({ photo = '' }) => (
  <Image
    style={styles.containerImage}
    source={{
      uri: photo.includes('http') ? photo : 'https://cdn4.iconfinder.com/data/icons/ionicons/512/icon-ios7-contact-512.png'
    }}
  />
);

const renderIcon = ({ onPress, name, size }) => (
  <TouchableOpacity
    onPress={onPress}
  >
    <Icon
      name={name}
      size={size}
    />
  </TouchableOpacity>
);

const renderInfo = ({ firstName, lastName, age }) => (
  <View>
    <Text style={styles.textSubject}>{`Name: ${firstName} ${lastName}`}</Text>
    <Text style={styles.textSubject}>{`Age: ${age} years old`}</Text>
  </View>
);

const renderContactInfo = ({ firstName, lastName, age, photo }) => (
  <View style={styles.row}>
    {renderImage({ photo })}
    {renderInfo({ firstName, lastName, age })}
  </View>
);

const renderButton = ({ onPress, color, children, last }) => (
  <View style={styles.containerFlex(last)}>
    <TouchableOpacity
      style={styles.button(color)}
      onPress={onPress}
      activeOpacity={1}
    >
      {children}
    </TouchableOpacity>
  </View>
);

const renderFormInput = ({ label, value, onChange, last, numeric }) => (
  <View style={styles.containerFlex(last)}>
    <Text style={styles.textBody}>{label}</Text>
    <TextInput
      keyboardType={numeric ? 'numeric' : 'default'}
      value={String(value)}
      onChangeText={onChange}
      style={styles.input}
    />
  </View>
);

const renderContact = ({ handlePressIcon, firstName, lastName, age, photo, display }) => (
  <View style={styles.container}>
    {renderContactInfo({ firstName, lastName, age, photo })}
    { display === 'none'
      ? renderIcon({ onPress: handlePressIcon(), name: 'ellipsis-horizontal-circle-outline', size: 24 })
      : renderIcon({ onPress: handlePressIcon(true), name: 'close-circle-outline', size: 24 })}
  </View>
);

const renderButtonEditDelete = ({ handleButtonEdit, handleDelete, display }) => (
  <View style={styles.contentContainer(display)}>
    <View style={styles.row}>
      {renderButton({
        children: <Text style={styles.buttonText}>Edit</Text>,
        onPress: handleButtonEdit
      })}
      {renderButton({
        color: '#eb2323',
        children: <Text style={styles.buttonText}>Delete</Text>,
        onPress: handleDelete,
        last: true
      })}
    </View>
  </View>
);

const renderButtonEditClose = ({ handleEdit, handleButtonClose, edit, loading, handleEditDone }) => (
  <View style={styles.row}>
    {renderButton({ 
      onPress: handleEdit,
      children: edit
        ? loading
        ? <ActivityIndicator
            size='small'
            color='#e4f9f5'
          />
        : handleEditDone()
        : <Text style={styles.buttonText}>Edit</Text>
    })}
    {renderButton({
      onPress: handleButtonClose,
      children: <Text style={styles.buttonText}>Close</Text>,
      color: '#eb2323',
      last: true
    })}
  </View>
);

const renderForm = ({
  display,
  firstName,
  lastName,
  age,
  setLastName,
  setFirstName,
  setAge,
  handleEdit,
  handleButtonClose,
  edit,
  loading,
  handleEditDone
}) => (
  <View style={styles.contentContainer(display)}>
    <View style={styles.row}>
      {renderFormInput({ label: 'First Name', value: firstName, onChange: firstName => setFirstName(firstName) })}
      {renderFormInput({ label: 'Last Name', value: lastName, onChange: lastName => setLastName(lastName) })}
      {renderFormInput({ label: 'Age', value: age, onChange: age => setAge(age), last: true, numeric: true })}
    </View>
    {renderButtonEditClose({ handleEdit, handleButtonClose, edit, loading, handleEditDone })}
  </View>
)

const Card = (props) => {
  const {
    age,
    edit,
    photo,
    bottom,
    detail,
    idCard,
    display,
    loading,
    lastName,
    firstName,
    setAge,
    handleEdit,
    setLastName,
    setFirstName,
    handleDelete,
    handlePressIcon,
    handleButtonEdit,
    handleButtonClose,
    handleEditDone
  } = useCard(props)

  return (
    <View style={{marginBottom: bottom}}>
      {renderContact({ ...props, handlePressIcon, display })}

      { detail && idCard === props.id
        ? renderButtonEditDelete({ handleDelete, handleButtonEdit, display })
        : renderForm({
          edit,
          loading,
          display,
          firstName,
          setFirstName,
          lastName,
          setLastName,
          age,
          setAge,
          handleEdit,
          handleButtonClose,
          handleEditDone
        })
      }
    </View>
  )
}

Card.propTypes = config.propTypes;
Card.displayName = config.displayName;

export default Card;
