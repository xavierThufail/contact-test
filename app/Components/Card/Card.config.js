import PropTypes from 'prop-types';

const propTypes = {
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  age: PropTypes.number.isRequired,
  photo: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired
};

const displayName = 'Card';

export default {
  propTypes,
  displayName
};
