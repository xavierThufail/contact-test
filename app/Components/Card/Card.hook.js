import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import allAction from '../../Store/actions';

const validateEdit = (editedValue) => {
  const error = [];
  if (editedValue.firstName.length < 4) {
    error.push('First Name must be 5 or more char');
  }
  if (editedValue.lastName.length < 4) {
    error.push('Last Name must be 5 or more char');
  }
  if (editedValue.age < 0) {
    error.push('Age is Empty');
  }

  return error;
};

const useCard = (props) => {
  const { idCard, loading, status } = useSelector(state => state.contact);


  const [edit, setEdit] = useState(false);
  const [display, setDisplay] = useState('none');
  const [detail, setDetail] = useState(false);
  const [bottom, setBottom] = useState(0);
  const [firstName, setFirstName] = useState(props.firstName);
  const [lastName, setLastName] = useState(props.lastName);
  const [age, setAge] = useState(String(props.age));
  const [photo, setPhoto] = React.useState(props.photo);

  const dispatch = useDispatch();

  const handleEdit = () => {
    const contactEdited = { firstName, lastName, age: Number(age), photo: props.photo ? props.photo : 'N/A' };
    const id = props.id;

    const error = validateEdit(contactEdited);

    if(error.length > 0) {
      dispatch(allAction.contact.setStatus(error[0]))
    } else {
      dispatch(allAction.contact.editContact(id, contactEdited));
      setEdit(true);
    }
  }

  const handleEditDone = () => {
    setEdit(false);
    handleButtonClose();
  }

  const handleDelete = () => {
    const id = props.id;
    dispatch(allAction.contact.deleteContact(id))
  }

  const handlePressIcon = (close) => () => {
    setDetail(close ? false : !detail);
    setDisplay(display == 'flex' ? 'none' : 'flex')
    setBottom(display == 'flex' ? 0 : 10)
    dispatch(allAction.contact.setIdCard(props.id));
  };

  const handleButtonClose = () => {
    setDisplay('none')
    setFirstName(props.firstName);
    setLastName(props.lastName);
    setAge(String(props.age));
  };

  const handleButtonEdit = () => {
    setDisplay('flex');
    setBottom(10);
    setDetail(false);
  };

  if (status) {
    setTimeout(() => {
      dispatch(allAction.contact.setStatus(''))
    }, 3000)
  }

  if (props.id !== idCard && (detail || display === 'flex')) {
    setDetail(false);
    setDisplay('none');
  }

  return {
    age,
    edit,
    photo,
    status,
    bottom,
    detail,
    idCard,
    display,
    loading,
    lastName,
    firstName,
    setAge,
    setEdit,
    handleEdit,
    setLastName,
    setFirstName,
    handleDelete,
    handlePressIcon,
    handleButtonEdit,
    handleButtonClose,
    handleEditDone
  }
};

export default useCard;
