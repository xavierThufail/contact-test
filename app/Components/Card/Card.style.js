const styles = {
  container: { 
    elevation: 10,
    flex: 1,
    backgroundColor: '#fff',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderRadius: 10,
    marginBottom: 10,
    padding: 10,
    paddingRight: 20,
    zIndex: 1
  },
  containerImage: {
    height: 50,
    width: 50,
    borderRadius:30,
  },
  contentContainer: (display) => ({
    backgroundColor: '#f0f0f0',
    padding: 20,
    paddingTop: 50,
    borderBottomLeftRadius: 10,
    borderBottomRightRadius: 10,
    elevation: 5,
    maxHeight: 350,
    marginTop: -50,
    display
  }),
  row: {
    flexDirection: 'row',
    alignItems: 'center', 
  },
  textSubject: {
    marginBottom: 10,
    marginLeft: 10,
    fontSize: 18,
    color: "black",
    fontWeight: 'bold'
  },
  buttonText: {
    textAlign: 'center',
    color: '#f0f0f0',
    fontWeight: 'bold'
  },
  button: (color) => ({
    minHeight: 30,
    minWidth: 50,
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: color ? color : '#1b83e3',
    padding: 10,
  }),
  input: {
    borderWidth: 1,
    height: 40,
    borderRadius: 10,
    marginBottom: 10,
    paddingLeft: 20
  },
  containerFlex: (last) => ({
    flex: 1,
    marginRight: last ? 0 : 10
  }),
  textBody: {
    fontWeight: 'bold'
  }
};

export default styles;
