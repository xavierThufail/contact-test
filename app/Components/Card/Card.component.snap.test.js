import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import Card from './Card.component';
import store from '../../Store';

jest
  .mock('../Icon', () => 'Icon')

describe('Card Component snapshot test', () => {
  it('Should render correctly', () => {
    const mockedData = {
      firstName: 'John',
      lastName: 'Doe',
      age: 90,
      photo: 'N/A',
      id: 10
    };

    const { toJSON } = render(
      <Provider store={store}>
        <Card {...mockedData} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot()
  })
});
