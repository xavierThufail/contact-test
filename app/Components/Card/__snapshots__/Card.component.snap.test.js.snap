// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component snapshot test Should render correctly 1`] = `
<View
  style={
    Object {
      "marginBottom": 0,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 10,
        "elevation": 10,
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "padding": 10,
        "paddingRight": 20,
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "https://cdn4.iconfinder.com/data/icons/ionicons/512/icon-ios7-contact-512.png",
          }
        }
        style={
          Object {
            "borderRadius": 30,
            "height": 50,
            "width": 50,
          }
        }
      />
      <View>
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginLeft": 10,
            }
          }
        >
          Name: John Doe
        </Text>
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginLeft": 10,
            }
          }
        >
          Age: 90 years old
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#1b83e3",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        ïŠ 
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#f0f0f0",
        "borderBottomLeftRadius": 10,
        "borderBottomRightRadius": 10,
        "display": "none",
        "elevation": 5,
        "marginTop": -50,
        "maxHeight": 350,
        "padding": 20,
        "paddingTop": 50,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          First Name
        </Text>
        <TextInput
          allowFontScaling={true}
          keyboardType="default"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          style={
            Object {
              "borderRadius": 10,
              "borderWidth": 1,
              "height": 40,
              "marginBottom": 10,
              "paddingLeft": 20,
            }
          }
          underlineColorAndroid="transparent"
          value="John"
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Last Name
        </Text>
        <TextInput
          allowFontScaling={true}
          keyboardType="default"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          style={
            Object {
              "borderRadius": 10,
              "borderWidth": 1,
              "height": 40,
              "marginBottom": 10,
              "paddingLeft": 20,
            }
          }
          underlineColorAndroid="transparent"
          value="Doe"
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 0,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Age
        </Text>
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          style={
            Object {
              "borderRadius": 10,
              "borderWidth": 1,
              "height": 40,
              "marginBottom": 10,
              "paddingLeft": 20,
            }
          }
          underlineColorAndroid="transparent"
          value="90"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 10,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#1b83e3",
              "borderRadius": 10,
              "justifyContent": "center",
              "minHeight": 30,
              "minWidth": 50,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#f0f0f0",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Edit
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 0,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#eb2323",
              "borderRadius": 10,
              "justifyContent": "center",
              "minHeight": 30,
              "minWidth": 50,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#f0f0f0",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Close
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
